p2panda-rs/src/document/document.rs

```diff
@@ -0,0 +1,390 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use std::convert::TryFrom;
+
+// // Now we use our own graph for topo sorting :-)
+// use incremental_topo::IncrementalTopo;
+
+use crate::document::{DocumentBuilderError, DocumentError};
+use crate::hash::Hash;
+use crate::identity::Author;
+use crate::instance::Instance;
+use crate::materialiser::Graph;
+use crate::operation::{AsOperation, OperationWithMeta};
+use crate::schema::Schema;
+
+/// Hard coded cddl string for now
+const DOCUMENT_SCHEMA: &str = "cafe = { (
+    name: { type: \"str\", value: tstr }
+) }";
+
+/// An iterator struct for Document.
+#[derive(Debug)]
+pub struct DocumentIter(Vec<OperationWithMeta>);
+
+/// A resolvable data type made up of a collection of causally linked operations.
+#[derive(Debug, Clone)]
+pub struct Document {
+    /// The create operation which is the root of this document.
+    create_operation: OperationWithMeta,
+    /// A causal graph of this documents operations which can be topologically sorted.
+    graph: Graph<OperationWithMeta>,
+}
+
+impl Document {
+    /// The hash id of this document.
+    pub fn id(&self) -> Hash {
+        self.create_operation.operation_id().to_owned()
+    }
+
+    /// The hash id of this documents schema.
+    pub fn schema_hash(&self) -> Hash {
+        self.create_operation.schema()
+    }
+
+    /// The author of this document.
+    pub fn author(&self) -> Author {
+        self.create_operation.public_key().to_owned()
+    }
+
+    /// The schema for this document.
+    pub fn schema(&self) -> Schema {
+        Schema::new(&self.schema_hash(), DOCUMENT_SCHEMA).unwrap()
+    }
```

Comment:
Maybe this should be a getter for `schema_hash` instead and we pass in the `schema_definition` in the `OperationBuilder.build()` method instead?